captum/attr/_core/shapley_sampling.py

```diff
@@ -0,0 +1,408 @@
+
+            # Iterate for number of samples, generate a permutation of the features
+            # and evalute the incremntal increase for each feature.
+            for i in range(n_samples):
+                feature_permutation = torch.randperm(total_features + 1)
+                prev_results = initial_eval
+                for (
+                    current_inputs,
+                    current_add_args,
+                    current_target,
+                    current_masks,
+                ) in self._ablation_generator(
+                    inputs,
+                    additional_forward_args,
+                    target,
+                    baselines,
+                    feature_mask,
+                    feature_permutation,
+                    ablations_per_eval,
+                ):
+                    # modified_eval dimensions: 1D tensor with length
+                    # equal to #num_examples * #features in batch
+                    modified_eval = _run_forward(
+                        self.forward_func,
+                        current_inputs,
+                        current_target,
+                        current_add_args,
+                    )
+                    # eval_diff dimensions: (#features in batch, #num_examples, 1,.. 1)
+                    # (contains 1 more dimension than inputs). This adds extra
+                    # dimensions of 1 to make the tensor broadcastable with the
+                    # inputs tensor.
+                    if single_output_mode:
+                        eval_diff = modified_eval - prev_results
+                        prev_results = modified_eval
+                    else:
+                        all_eval = torch.cat((prev_results, modified_eval), dim=0)
+                        eval_diff = all_eval[num_examples:] - all_eval[:-num_examples]
+                        prev_results = all_eval[-num_examples:]
+                    for j in range(len(total_attrib)):
+                        current_eval_diff = eval_diff
+                        if not single_output_mode:
+                            current_eval_diff = current_eval_diff.reshape(
+                                (-1, num_examples) + (len(inputs[j].shape) - 1) * (1,)
+                            )
+                        total_attrib[j] += (
+                            current_eval_diff * current_masks[j].float()
+                        ).sum(dim=0)
+
+            # Divide total attributions by number of random permutations and return
+            # formatted attributions.
+            attrib = tuple(
+                tensor_attrib_total / n_samples for tensor_attrib_total in total_attrib
+            )
+            return _format_attributions(is_inputs_tuple, attrib)
+
+    def _ablation_generator(
+        self,
+        inputs,
+        additional_args,
+        target,
+        baselines,
+        input_masks,
+        feature_permutation,
+        ablations_per_eval,
+    ):
+        current_tensors = baselines
```

Comment:
I think naming it current_baselines might be confusing as well, since it's the same variable in which we add each feature. I'll add a comment to make the naming clearer?